encfs/MACFileIO.cpp

```diff
@@ -137,78 +141,125 @@ off_t MACFileIO::getSize() const {
-          cipher->MAC_64(tmp.data + macBytes, readSize - macBytes, key);
-
-      // Constant time comparision to prevent timing attacks
-      unsigned char fail = 0;
-      for (int i = 0; i < macBytes; ++i, mac >>= 8) {
-        int test = mac & 0xff;
-        int stored = tmp.data[i];
-
-        fail |= (test ^ stored);
-      }
+    IORequest tmp;
+    tmp.offset = locWithHeader(req.offset, bs, headerSize);
+    tmp.data = mb.data;
+    tmp.dataLen = headerSize + req.dataLen;
 
-      if (fail > 0) {
-        // uh oh..
-        long blockNum = req.offset / bs;
-        RLOG(WARNING) << "MAC comparison failure in block " << blockNum;
-        if (!warnOnly) {
-          MemoryPool::release(mb);
-          throw Error(_("MAC comparison failure, refusing to read"));
+    // get the data from the base FileIO layer
+    ssize_t readSize = base->read(tmp);
+
+    // don't store zeros if configured for zero-block pass-through
+    bool skipBlock = true;
+    if (_allowHoles) {
+      for (int i = 0; i < readSize; ++i)
+        if (tmp.data[i] != 0) {
+          skipBlock = false;
+          break;
+        }
+    } else if (macBytes > 0)
+      skipBlock = false;
+
+    if (readSize > headerSize) {
+      if (!skipBlock) {
+        // At this point the data has been decoded.  So, compute the MAC of
+        // the block and check against the checksum stored in the header..
+        uint64_t mac =
+            cipher->MAC_64(tmp.data + macBytes, readSize - macBytes, key);
+
+        // Constant time comparision to prevent timing attacks
+        unsigned char fail = 0;
+        for (int i = 0; i < macBytes; ++i, mac >>= 8) {
+          int test = mac & 0xff;
+          int stored = tmp.data[i];
+
+          fail |= (test ^ stored);
         }
+
+        if (fail > 0) {
+          // uh oh..
+          long blockNum = req.offset / bs;
+          RLOG(WARNING) << "MAC comparison failure in block " << blockNum;
+          if (!warnOnly) {
+            MemoryPool::release(mb);
+            throw Error(_("MAC comparison failure, refusing to read"));
+          }
+        }
+      }
+
+      // now copy the data to the output buffer
+      readSize -= headerSize;
+      memcpy(req.data, tmp.data + headerSize, readSize);
+    } else {
+      VLOG(1) << "readSize " << readSize << " at offset " << req.offset;
+      if (readSize > 0) readSize = 0;
+    }
+
+    MemoryPool::release(mb);
+
+    return readSize;
+
+  } else {	  //reverse = true
+
+    if (randBytes > 0) {
+      throw Error(_("MAC rand bytes not yet implemented in reverse mode"));
+      return -1;
+    }
+
+    MemBlock mb = MemoryPool::allocate(blockSize());
+
+    IORequest tmp;
+    tmp.offset = locWithoutHeader(req.offset, blockSize(), macBytes);
+    tmp.data = mb.data + macBytes;
+    tmp.dataLen = req.dataLen - macBytes;
+
+    // get the data from the base FileIO layer
+    ssize_t readSize = base->read(tmp);
+    if (readSize == 0) {
+      MemoryPool::release(mb);
+      return 0;
```

Comment:
Code below will fail if `readSize < 0`.
It could occur in case of storage IO error.
In this case you should return `readSize` (as done above).
Thx !